prow/cmd/pipeline/controller.go

```diff
@@ -0,0 +1,673 @@
+	eventBroadcaster.StartLogging(logrus.Infof)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kc.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, untypedcorev1.EventSource{Component: controllerName})
+
+	c := &controller{
+		config:     prowConfig,
+		pjc:        pjc,
+		pipelines:  pipelineConfigs,
+		pjLister:   pji.Lister(),
+		pjInformer: pji.Informer(),
+		workqueue:  rl,
+		recorder:   recorder,
+		totURL:     totURL,
+	}
+
+	logrus.Info("Setting up event handlers")
+
+	// Reconcile whenever a prowjob changes
+	pji.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pj, ok := obj.(*prowjobv1.ProwJob)
+			if !ok {
+				logrus.Warnf("Ignoring bad prowjob add: %v", obj)
+				return
+			}
+			c.enqueueKey(pj.Spec.Cluster, pj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			pj, ok := new.(*prowjobv1.ProwJob)
+			if !ok {
+				logrus.Warnf("Ignoring bad prowjob update: %v", new)
+				return
+			}
+			c.enqueueKey(pj.Spec.Cluster, pj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pj, ok := obj.(*prowjobv1.ProwJob)
+			if !ok {
+				logrus.Warnf("Ignoring bad prowjob delete: %v", obj)
+				return
+			}
+			c.enqueueKey(pj.Spec.Cluster, pj)
+		},
+	})
+
+	for ctx, cfg := range pipelineConfigs {
+		// Reconcile whenever a pipelinerun changes.
+		ctx := ctx // otherwise it will change
+		cfg.informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueKey(ctx, obj)
+			},
+			UpdateFunc: func(old, new interface{}) {
+				c.enqueueKey(ctx, new)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueKey(ctx, obj)
+			},
+		})
+	}
+
+	return c, nil
+}
+
+// Run starts threads workers, returning after receiving a stop signal.
+func (c *controller) Run(threads int, stop <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	logrus.Info("Starting Pipeline controller")
+	logrus.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stop, c.hasSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	logrus.Info("Starting workers")
+	for i := 0; i < threads; i++ {
+		go wait.Until(c.runWorker, time.Second, stop)
+	}
+
+	logrus.Info("Started workers")
+	<-stop
+	logrus.Info("Shutting down workers")
+	return nil
+}
+
+// runWorker dequeues to reconcile, until the queue has closed.
+func (c *controller) runWorker() {
+	for {
+		key, shutdown := c.workqueue.Get()
+		if shutdown {
+			return
+		}
+		func() {
+			defer c.workqueue.Done(key)
+
+			if err := reconcile(c, key.(string)); err != nil {
+				runtime.HandleError(fmt.Errorf("failed to reconcile %s: %v", key, err))
+				return // Do not forget so we retry later.
+			}
+			c.workqueue.Forget(key)
+		}()
+	}
+}
+
+// toKey returns context/namespace/name/kind
+func toKey(ctx, namespace, name, kind string) string {
```

Comment:
Instead of including the kind in the key so that we can distinguish ProwJobs from Pipelines can we instead always name the Pipeline with the same name as the ProwJob and then look up both the ProwJob and the Pipeline when reconciling? This is what we do in the build controller: https://github.com/kubernetes/test-infra/blob/b7b7a8907608cc78abae88860367eb6710157aa5/prow/cmd/build/controller.go#L397-L414
Plank also creates pods with names matching the owning ProwJob. This is very helpful for debugging and is assumed in a number of places so we should try to avoid breaking the assumption.

Reply:
The PipelineRun is created with the same name as the ProwJob, but it has has a label with the ProwJob ID, which we used as a selector.

The reason for these is that in Jenkins X, we are using a different name for the pipeline run that encodes the repository + branch names, that is not the case anymore for raw pipeline use case.

If you consider that this change is really necessary, I can refactor the code.

Reply:
Yes please lets keep this identical between Prow controllers

Reply:
I removed the `kind` from the key and refactored the code to follow the same logic like the build controller.